contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,621 @@
+           uint amountInputToken =  _transferShortfallFromSender(_longTokenUnderlying, _amountRequired, longTokenObtained, originalSender, paymentParams);
+           emit ExchangeIssue(originalSender, ISetToken(setToken), IERC20(_longTokenUnderlying), amountInputToken, setAmount);
+        }
+        else {
+            // TODO: Add Implementation of other payment options
+            revert("Not Implemented");
+        }
+    }
+
+    function _issueSet(bytes memory _params) internal {
+        (address setToken, uint256 setAmount, address originalSender,,,) = _decodeParams(_params);
+        debtIssuanceModule.issue(ISetToken(setToken), setAmount, originalSender);
+    }
+
+    function _decodeParams(bytes memory params) internal pure returns(address setToken, uint256 setAmount, address originalSender, bool isIssuance, PaymentToken paymentToken, bytes memory paymentParams){
+            (setToken, setAmount, originalSender, isIssuance, paymentToken, paymentParams) = abi.decode(params, (address, uint256, address, bool, PaymentToken, bytes));
+    }
+
+    function _decodePaymentParamsLongToken(bytes memory _paymentParams) internal pure returns(uint256 maxAmountInputToken){
+            maxAmountInputToken = abi.decode(_paymentParams, (uint256));
+    }
+
+
+    /**
+     * Transfers the shortfall between the amount of tokens required to return flashloan and what was obtained
+     * from swapping the debt tokens from the users address
+     *
+     * @param _token                 Address of the token to transfer from user
+     * @param _amountRequired        Amount required to repay flashloan
+     * @param _amountObtained        Amount obtained from swapping the short token
+     * @param _originalSender        Adress that initiated the token issuance, which is the adresss form which to transfer the tokens
+     */
+    function _transferShortfallFromSender(address _token, uint256 _amountRequired, uint256 _amountObtained, address _originalSender, bytes memory paymentParams) internal returns(uint256 shortfall) {
+        if(_amountObtained < _amountRequired){ 
+            shortfall = _amountRequired.sub(_amountObtained);
+            uint256 maxAmountInputToken =_decodePaymentParamsLongToken(paymentParams);
+            require(shortfall <= maxAmountInputToken, "ExchangeIssuance: INSUFFICIENT INPUT AMOUNT");
+            IERC20(_token).safeTransferFrom(_originalSender, address(this), shortfall);
+        }
+    }
+
+
+
+    /**
+     * Swaps the debt tokens obtained from issuance for the underlying of the collateral
+     *
+     * @param _setToken                   Address of the SetToken to be issued
+     * @param _setAmount                  Amount of SetTokens to issue
+     * @param _longTokenUnderlying        Address of the underlying of the collateral token
+     */
+    function _swapShortForLongTokenUnderlying(address _setToken, uint256 _setAmount, address _longTokenUnderlying) internal returns (uint256) {
```

Comment:
Can we instead allow the user to select what exchange to use using an enum? This removes the gas costs of needing to figure out the best exchange on-chain, and lets us add new exchanges much more easily.